modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.ignite.internal.util.collection.IntHashMap;
+import org.apache.ignite.internal.util.collection.IntMap;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+import static org.apache.ignite.internal.processors.cache.GridCacheUtils.EXPIRE_TIME_CALCULATE;
+import static org.apache.ignite.internal.processors.cache.GridCacheUtils.TTL_NOT_CHANGED;
+
+/**
+ * Contains logic to process {@link CdcEvent} and apply them to the destination cluster.
+ *
+ * @see IgniteInternalCache#putAllConflict(Map)
+ * @see IgniteInternalCache#removeAllConflict(Map)
+ */
+public class CdcEventsIgniteApplier extends CdcEventsApplier<KeyCacheObject, GridCacheDrInfo> {
+    /** Destination cluster. */
+    private final IgniteEx ignite;
+
+    /** Caches. */
+    private final IntMap<IgniteInternalCache<BinaryObject, BinaryObject>> ignCaches = new IntHashMap<>();
+
+    /**
+     * @param ignite Destination cluster.
+     * @param maxBatchSize Maximum batch size.
+     * @param log Logger.
+     */
+    public CdcEventsIgniteApplier(Ignite ignite, int maxBatchSize, IgniteLogger log) {
+        super(maxBatchSize, log);
+
+        this.ignite = (IgniteEx)ignite;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void putAllConflict(int cacheId, Map<KeyCacheObject, GridCacheDrInfo> drMap) {
+        try {
+            IgniteInternalCache<BinaryObject, BinaryObject> cache = cache(cacheId);
+
+            if (cache.configuration().getExpiryPolicyFactory() != null) {
+                drMap = F.viewReadOnly(drMap,
+                    (info) -> new GridCacheDrExpirationInfo(info.value(), info.version(),
+                        TTL_NOT_CHANGED, EXPIRE_TIME_CALCULATE));
+            }
+
+            cache(cacheId).putAllConflict(drMap);
+        }
+        catch (IgniteCheckedException e) {
+            throw U.convertException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void removeAllConflict(int cacheId, Map<KeyCacheObject, GridCacheVersion> drMap) {
+        try {
+            cache(cacheId).removeAllConflict(drMap);
+        }
+        catch (IgniteCheckedException e) {
+            throw U.convertException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected KeyCacheObject toKey(CdcEvent evt) {
+        Object key = evt.key();
+
+        if (key instanceof KeyCacheObject)
+            return (KeyCacheObject)key;
+        else
+            return new KeyCacheObjectImpl(key, null, evt.partition());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected GridCacheDrInfo toValue(Object val, GridCacheVersion ver) {
+        CacheObject cacheObj;
+
+        if (val instanceof CacheObject)
+            cacheObj = (CacheObject)val;
+        else
+            cacheObj = new CacheObjectImpl(val, null);
+
+        return new GridCacheDrInfo(cacheObj, ver);
```

Comment:
Let's check here expiry policy for cache to eliminate object recreation on apply.